File: 05-oidc/typescript/provider.ts
import crypto from "node:crypto";
import http, { IncomingMessage, ServerResponse } from "node:http";
import { randomUUID } from "node:crypto";

const ISSUER = "http://localhost:8000";
const ACCESS_TTL = 3600;
const ID_TOKEN_TTL = 3600;
const REFRESH_TTL = 86400 * 7;
const AUTH_CODE_TTL = 300;

const { publicKey, privateKey } = crypto.generateKeyPairSync("rsa", {
  modulusLength: 2048,
  publicKeyEncoding: { type: "spki", format: "pem" },
  privateKeyEncoding: { type: "pkcs8", format: "pem" },
});

const USERS: Record<string, any> = {
  alice: {
    password: process.env.ALICE_PASSWORD ?? "password-alice",
    sub: "user-alice-001",
    name: "Alice Johnson",
    given_name: "Alice",
    family_name: "Johnson",
    email: "alice@example.com",
    email_verified: true,
    picture: "https://example.com/avatars/alice.jpg",
  },
  bob: {
    password: process.env.BOB_PASSWORD ?? "password-bob",
    sub: "user-bob-002",
    name: "Bob Smith",
    given_name: "Bob",
    family_name: "Smith",
    email: "bob@example.com",
    email_verified: false,
    picture: "https://example.com/avatars/bob.jpg",
  },
};

const CLIENTS: Record<string, any> = {
  rp: {
    client_secret: process.env.RP_SECRET ?? "rp-secret",
    redirect_uris: ["http://localhost:8001/callback"],
    grant_types: ["authorization_code", "refresh_token"],
  },
  spa: {
    client_secret: null,
    redirect_uris: ["http://localhost:3000/callback"],
    grant_types: ["authorization_code"],
  },
};

const authCodes = new Map<string, any>();
const refreshTokens = new Map<string, any>();

function base64url(buf: Buffer): string { return buf.toString("base64url"); }
function randomToken(): string { return crypto.randomBytes(48).toString("base64url"); }

function encodeJwt(payload: Record<string, unknown>, key: string): string {
  const header = { alg: "RS256", typ: "JWT", kid: "oidc-rsa-1" };
  const h = base64url(Buffer.from(JSON.stringify(header)));
  const p = base64url(Buffer.from(JSON.stringify(payload)));
  const sig = crypto.sign("sha256", Buffer.from(`${h}.${p}`), key);
  return `${h}.${p}.${base64url(sig)}`;
}

function makeIdToken(user: any, clientId: string, nonce: string | null, authTime: number): string {
  const now = Math.floor(Date.now() / 1000);
  const claims: Record<string, any> = {
    iss: ISSUER, sub: user.sub, aud: [clientId],
    exp: now + ID_TOKEN_TTL, iat: now, auth_time: authTime,
    nonce, azp: clientId,
  };
  for (const c of ["name", "given_name", "family_name", "email", "email_verified", "picture"]) {
    if (user[c] !== undefined) claims[c] = user[c];
  }
  return encodeJwt(claims, privateKey);
}

function makeAccessToken(user: any, scope: string, clientId: string): string {
  const now = Math.floor(Date.now() / 1000);
  return encodeJwt({ iss: ISSUER, sub: user.sub, client_id: clientId, scope, iat: now, exp: now + ACCESS_TTL, jti: randomUUID() }, privateKey);
}

function extractScopeClaims(scope: string, user: any): Record<string, any> {
  const result: Record<string, any> = { sub: user.sub };
  const scopes = scope.split(" ");
  if (scopes.includes("profile")) {
    for (const c of ["name", "given_name", "family_name", "picture"]) {
      if (user[c] !== undefined) result[c] = user[c];
    }
  }
  if (scopes.includes("email")) {
    for (const c of ["email", "email_verified"]) {
      if (user[c] !== undefined) result[c] = user[c];
    }
  }
  return result;
}

function findUserBySub(sub: string): any {
  for (const u of Object.values(USERS)) if (u.sub === sub) return u;
  return null;
}

function readBody(req: IncomingMessage): Promise<string> {
  return new Promise((resolve) => {
    const chunks: Buffer[] = [];
    req.on("data", (c) => chunks.push(c));
    req.on("end", () => resolve(Buffer.concat(chunks).toString()));
  });
}

function sendJson(res: ServerResponse, status: number, data: any) {
  res.writeHead(status, { "Content-Type": "application/json" });
  res.end(JSON.stringify(data));
}

function sendHtml(res: ServerResponse, html: string) {
  res.writeHead(200, { "Content-Type": "text/html" });
  res.end(html);
}

function sendRedirect(res: ServerResponse, loc: string) {
  res.writeHead(302, { Location: loc });
  res.end();
}

function parseForm(body: string): Record<string, string> {
  const r: Record<string, string> = {};
  for (const p of body.split("&")) {
    const [k, ...v] = p.split("=");
    r[decodeURIComponent(k)] = decodeURIComponent(v.join("="));
  }
  return r;
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url ?? "/", `http://${req.headers.host}`);
  const path = url.pathname;
  const method = req.method ?? "GET";
  const q = Object.fromEntries(url.searchParams);

  try {
    if (path === "/authorize" && method === "GET") {
      const client = CLIENTS[q.client_id];
      if (!client || !client.redirect_uris.includes(q.redirect_uri)) {
        sendJson(res, 400, { error: "invalid_request" });
        return;
      }
      if (q.response_type !== "code") {
        sendJson(res, 400, { error: "unsupported_response_type" });
        return;
      }
      if (!q.scope?.split(" ").includes("openid")) {
        sendJson(res, 400, { error: "openid scope required" });
        return;
      }
      sendHtml(res, `<!DOCTYPE html>
<html><body style="font-family:sans-serif;max-width:500px;margin:40px auto">
<h2>Sign in</h2>
<form method="post" action="/consent">
<input type="hidden" name="response_type" value="${q.response_type}">
<input type="hidden" name="client_id" value="${q.client_id}">
<input type="hidden" name="redirect_uri" value="${q.redirect_uri}">
<input type="hidden" name="scope" value="${q.scope}">
<input type="hidden" name="state" value="${q.state ?? ""}">
<input type="hidden" name="nonce" value="${q.nonce ?? ""}">
<p><label>Username: <input name="username" value="alice"></label></p>
<p><label>Password: <input name="password" type="password"></label></p>
<p><button type="submit" name="approve" value="yes">Sign In</button>
<button type="submit" name="approve" value="no">Cancel</button></p>
</form></body></html>`);
      return;
    }

    if (path === "/consent" && method === "POST") {
      const body = await readBody(req);
      const f = parseForm(body);
      if (f.approve !== "yes") { sendJson(res, 403, { error: "access_denied" }); return; }

      const user = USERS[f.username];
      if (!user || user.password !== f.password) {
        sendJson(res, 401, { error: "invalid_credentials" });
        return;
      }

      const code = randomToken();
      const authTime = Math.floor(Date.now() / 1000);
      authCodes.set(code, {
        client_id: f.client_id, redirect_uri: f.redirect_uri,
        scope: f.scope, nonce: f.nonce || null,
        username: f.username, auth_time: authTime,
        expires: Date.now() + AUTH_CODE_TTL * 1000,
      });

      const p = new URLSearchParams({ code });
      if (f.state) p.set("state", f.state);
      sendRedirect(res, `${f.redirect_uri}?${p}`);
      return;
    }

    if (path === "/token" && method === "POST") {
      const body = await readBody(req);
      const f = parseForm(body);
      if (f.grant_type === "authorization_code") return handleAuthCode(res, f);
      if (f.grant_type === "refresh_token") return handleRefresh(res, f);
      sendJson(res, 400, { error: "unsupported_grant_type" });
      return;
    }

    if (path === "/userinfo" && method === "GET") {
      const auth = req.headers.authorization ?? "";
      if (!auth.startsWith("Bearer ")) {
        sendJson(res, 401, { error: "missing_token" });
        return;
      }
      const token = auth.slice(7);
      const parts = token.split(".");
      if (parts.length !== 3) { sendJson(res, 401, { error: "invalid_token" }); return; }
      try {
        const valid = crypto.verify("sha256", Buffer.from(`${parts[0]}.${parts[1]}`), publicKey, Buffer.from(parts[2], "base64url"));
        if (!valid) { sendJson(res, 401, { error: "invalid_token" }); return; }
        const payload = JSON.parse(Buffer.from(parts[1], "base64url").toString());
        const user = findUserBySub(payload.sub);
        if (!user) { sendJson(res, 404, { error: "user_not_found" }); return; }
        sendJson(res, 200, extractScopeClaims(payload.scope ?? "", user));
      } catch {
        sendJson(res, 401, { error: "invalid_token" });
      }
      return;
    }

    if (path === "/.well-known/openid-configuration") {
      sendJson(res, 200, {
        issuer: ISSUER, authorization_endpoint: `${ISSUER}/authorize`,
        token_endpoint: `${ISSUER}/token`, userinfo_endpoint: `${ISSUER}/userinfo`,
        jwks_uri: `${ISSUER}/.well-known/jwks.json`,
        scopes_supported: ["openid", "profile", "email"],
        response_types_supported: ["code"],
        grant_types_supported: ["authorization_code", "refresh_token"],
        id_token_signing_alg_values_supported: ["RS256"],
      });
      return;
    }

    if (path === "/.well-known/jwks.json") {
      const pub = crypto.createPublicKey(publicKey);
      const jwk = pub.export({ format: "jwk" });
      sendJson(res, 200, { keys: [{ kty: jwk.kty, use: "sig", alg: "RS256", kid: "oidc-rsa-1", n: jwk.n, e: jwk.e }] });
      return;
    }

    sendJson(res, 404, { error: "not_found" });
  } catch {
    sendJson(res, 500, { error: "server_error" });
  }
});

function handleAuthCode(res: ServerResponse, f: Record<string, string>) {
  const client = CLIENTS[f.client_id];
  if (!client || (client.client_secret && client.client_secret !== f.client_secret)) {
    sendJson(res, 400, { error: "invalid_client" });
    return;
  }
  const stored = authCodes.get(f.code);
  if (!stored || stored.expires < Date.now() || stored.client_id !== f.client_id || stored.redirect_uri !== f.redirect_uri) {
    authCodes.delete(f.code);
    sendJson(res, 400, { error: "invalid_grant" });
    return;
  }
  authCodes.delete(f.code);

  const user = USERS[stored.username];
  const scope = stored.scope;
  const accessToken = makeAccessToken(user, scope, f.client_id);
  const idToken = makeIdToken(user, f.client_id, stored.nonce, stored.auth_time);
  const refreshToken = randomToken();
  refreshTokens.set(refreshToken, { client_id: f.client_id, username: stored.username, scope, expires: Date.now() + REFRESH_TTL * 1000 });

  sendJson(res, 200, { access_token: accessToken, token_type: "Bearer", expires_in: ACCESS_TTL, refresh_token: refreshToken, id_token: idToken, scope });
}

function handleRefresh(res: ServerResponse, f: Record<string, string>) {
  const stored = refreshTokens.get(f.refresh_token);
  if (!stored || stored.expires < Date.now() || stored.client_id !== f.client_id) {
    refreshTokens.delete(f.refresh_token);
    sendJson(res, 400, { error: "invalid_grant" });
    return;
  }
  refreshTokens.delete(f.refresh_token);
  const user = USERS[stored.username];
  const newAccess = makeAccessToken(user, stored.scope, f.client_id);
  const newRefresh = randomToken();
  const idToken = makeIdToken(user, f.client_id, null, Math.floor(Date.now() / 1000));
  refreshTokens.set(newRefresh, { client_id: f.client_id, username: stored.username, scope: stored.scope, expires: Date.now() + REFRESH_TTL * 1000 });
  sendJson(res, 200, { access_token: newAccess, token_type: "Bearer", expires_in: ACCESS_TTL, refresh_token: newRefresh, id_token: idToken, scope: stored.scope });
}

const PORT = parseInt(process.env.PORT ?? "8000", 10);
server.listen(PORT, "0.0.0.0", () => console.log(`OIDC Provider at http://localhost:${PORT}`));

File: 05-oidc/go/provider.go
package main

import (
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

var (
	issuer      = "http://localhost:8000"
	accessTTL   = 3600
	idTokenTTL  = 3600
	refreshTTL  = 86400 * 7
	authCodeTTL = 300
)

var rsaPrivateKey *rsa.PrivateKey
var rsaPublicKey *rsa.PublicKey

func init() {
	var err error
	rsaPrivateKey, err = rsa.GenerateKey(rand.Reader, 2048)
	if err != nil {
		log.Fatalf("Failed to generate RSA key: %v", err)
	}
	rsaPublicKey = &rsaPrivateKey.PublicKey
}

type userProfile struct {
	Password       string
	Sub            string
	Name           string
	GivenName      string
	FamilyName     string
	Email          string
	EmailVerified  bool
	Picture        string
}

var users = map[string]*userProfile{
	"alice": {
		Password:      getEnv("ALICE_PASSWORD", "password-alice"),
		Sub:           "user-alice-001",
		Name:          "Alice Johnson",
		GivenName:     "Alice",
		FamilyName:    "Johnson",
		Email:         "alice@example.com",
		EmailVerified: true,
		Picture:       "https://example.com/avatars/alice.jpg",
	},
	"bob": {
		Password:      getEnv("BOB_PASSWORD", "password-bob"),
		Sub:           "user-bob-002",
		Name:          "Bob Smith",
		GivenName:     "Bob",
		FamilyName:    "Smith",
		Email:         "bob@example.com",
		EmailVerified: false,
		Picture:       "https://example.com/avatars/bob.jpg",
	},
}

type client struct {
	ClientSecret string
	RedirectURIs []string
	GrantTypes   []string
}

var clients = map[string]*client{
	"rp":  {ClientSecret: getEnv("RP_SECRET", "rp-secret"), RedirectURIs: []string{"http://localhost:8001/callback"}, GrantTypes: []string{"authorization_code", "refresh_token"}},
	"spa": {RedirectURIs: []string{"http://localhost:3000/callback"}, GrantTypes: []string{"authorization_code"}},
}

type authCodeStore struct {
	ClientID    string
	RedirectURI string
	Scope       string
	Nonce       string
	Username    string
	AuthTime    int64
	Expires     time.Time
}

type refreshEntry struct {
	ClientID string
	Username string
	Scope    string
	Expires  time.Time
}

var (
	authCodes    = make(map[string]*authCodeStore)
	refreshStore = make(map[string]*refreshEntry)
	mu           sync.RWMutex
)

func getEnv(k, f string) string {
	if v := os.Getenv(k); v != "" {
		return v
	}
	return f
}

func randomToken() string {
	b := make([]byte, 36)
	rand.Read(b)
	return base64.RawURLEncoding.EncodeToString(b)
}

func makeIDToken(user *userProfile, clientID, nonce string, authTime int64) string {
	now := time.Now()
	claims := jwt.MapClaims{
		"iss": issuer, "sub": user.Sub, "aud": []string{clientID},
		"exp": now.Add(time.Duration(idTokenTTL) * time.Second).Unix(), "iat": now.Unix(),
		"auth_time": authTime, "azp": clientID, "name": user.Name,
		"given_name": user.GivenName, "family_name": user.FamilyName,
		"email": user.Email, "email_verified": user.EmailVerified, "picture": user.Picture,
	}
	if nonce != "" {
		claims["nonce"] = nonce
	}
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	token.Header["kid"] = "oidc-rsa-1"
	signed, _ := token.SignedString(rsaPrivateKey)
	return signed
}

func makeAccessToken(user *userProfile, scope, clientID string) string {
	now := time.Now()
	claims := jwt.MapClaims{
		"iss": issuer, "sub": user.Sub, "client_id": clientID, "scope": scope,
		"iat": now.Unix(), "exp": now.Add(time.Duration(accessTTL) * time.Second).Unix(),
	}
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, claims)
	signed, _ := token.SignedString(rsaPrivateKey)
	return signed
}

func extractScopeClaims(scope string, user *userProfile) map[string]any {
	r := map[string]any{"sub": user.Sub}
	if strings.Contains(scope, "profile") {
		r["name"] = user.Name
		r["given_name"] = user.GivenName
		r["family_name"] = user.FamilyName
		r["picture"] = user.Picture
	}
	if strings.Contains(scope, "email") {
		r["email"] = user.Email
		r["email_verified"] = user.EmailVerified
	}
	return r
}

func sendJSON(w http.ResponseWriter, s int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(s)
	json.NewEncoder(w).Encode(v)
}

func sendHTML(w http.ResponseWriter, h string) {
	w.Header().Set("Content-Type", "text/html")
	w.Write([]byte(h))
}

func authorizeHandler(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	cl, ok := clients[q.Get("client_id")]
	if !ok {
		sendJSON(w, 400, map[string]string{"error": "invalid_client"})
		return
	}
	validURI := false
	for _, u := range cl.RedirectURIs {
		if u == q.Get("redirect_uri") {
			validURI = true
			break
		}
	}
	if !validURI || q.Get("response_type") != "code" {
		sendJSON(w, 400, map[string]string{"error": "invalid_request"})
		return
	}
	if !strings.Contains(q.Get("scope"), "openid") {
		sendJSON(w, 400, map[string]string{"error": "openid_scope_required"})
		return
	}
	sendHTML(w, fmt.Sprintf(`<!DOCTYPE html>
<html><body style="font-family:sans-serif;max-width:500px;margin:40px auto">
<h2>Sign in</h2>
<form method="post" action="/consent">
<input type="hidden" name="response_type" value="%s">
<input type="hidden" name="client_id" value="%s">
<input type="hidden" name="redirect_uri" value="%s">
<input type="hidden" name="scope" value="%s">
<input type="hidden" name="state" value="%s">
<input type="hidden" name="nonce" value="%s">
<p><label>Username: <input name="username" value="alice"></label></p>
<p><label>Password: <input name="password" type="password"></label></p>
<p><button type="submit" name="approve" value="yes">Sign In</button>
<button type="submit" name="approve" value="no">Cancel</button></p>
</form></body></html>`,
		q.Get("response_type"), q.Get("client_id"), q.Get("redirect_uri"),
		q.Get("scope"), q.Get("state"), q.Get("nonce")))
}

func consentHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		sendJSON(w, 405, map[string]string{"error": "method_not_allowed"})
		return
	}
	r.ParseForm()
	if r.FormValue("approve") != "yes" {
		sendJSON(w, 403, map[string]string{"error": "access_denied"})
		return
	}
	user, ok := users[r.FormValue("username")]
	if !ok || user.Password != r.FormValue("password") {
		sendJSON(w, 401, map[string]string{"error": "invalid_credentials"})
		return
	}
	code := randomToken()
	authTime := time.Now().Unix()
	mu.Lock()
	authCodes[code] = &authCodeStore{
		ClientID: r.FormValue("client_id"), RedirectURI: r.FormValue("redirect_uri"),
		Scope: r.FormValue("scope"), Nonce: r.FormValue("nonce"),
		Username: r.FormValue("username"), AuthTime: authTime,
		Expires: time.Now().Add(time.Duration(authCodeTTL) * time.Second),
	}
	mu.Unlock()
	v := url.Values{"code": {code}}
	if s := r.FormValue("state"); s != "" {
		v.Set("state", s)
	}
	http.Redirect(w, r, r.FormValue("redirect_uri")+"?"+v.Encode(), http.StatusFound)
}

func tokenHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		sendJSON(w, 405, map[string]string{"error": "method_not_allowed"})
		return
	}
	r.ParseForm()
	switch r.FormValue("grant_type") {
	case "authorization_code":
		handleAuthCode(w, r)
	case "refresh_token":
		handleRefresh(w, r)
	default:
		sendJSON(w, 400, map[string]string{"error": "unsupported_grant_type"})
	}
}

func handleAuthCode(w http.ResponseWriter, r *http.Request) {
	cl, ok := clients[r.FormValue("client_id")]
	if !ok || (cl.ClientSecret != "" && cl.ClientSecret != r.FormValue("client_secret")) {
		sendJSON(w, 400, map[string]string{"error": "invalid_client"})
		return
	}
	mu.Lock()
	stored, exists := authCodes[r.FormValue("code")]
	if exists {
		delete(authCodes, r.FormValue("code"))
	}
	mu.Unlock()
	if !exists || time.Now().After(stored.Expires) ||
		stored.ClientID != r.FormValue("client_id") ||
		stored.RedirectURI != r.FormValue("redirect_uri") {
		sendJSON(w, 400, map[string]string{"error": "invalid_grant"})
		return
	}
	user, _ := users[stored.Username]
	scope := stored.Scope
	at := makeAccessToken(user, scope, r.FormValue("client_id"))
	id := makeIDToken(user, r.FormValue("client_id"), stored.Nonce, stored.AuthTime)
	rt := randomToken()
	mu.Lock()
	refreshStore[rt] = &refreshEntry{ClientID: r.FormValue("client_id"), Username: stored.Username, Scope: scope, Expires: time.Now().Add(time.Duration(refreshTTL) * time.Second)}
	mu.Unlock()
	sendJSON(w, 200, map[string]any{"access_token": at, "token_type": "Bearer", "expires_in": accessTTL, "refresh_token": rt, "id_token": id, "scope": scope})
}

func handleRefresh(w http.ResponseWriter, r *http.Request) {
	mu.Lock()
	stored, exists := refreshStore[r.FormValue("refresh_token")]
	if exists {
		delete(refreshStore, r.FormValue("refresh_token"))
	}
	mu.Unlock()
	if !exists || time.Now().After(stored.Expires) || stored.ClientID != r.FormValue("client_id") {
		sendJSON(w, 400, map[string]string{"error": "invalid_grant"})
		return
	}
	user, _ := users[stored.Username]
	at := makeAccessToken(user, stored.Scope, r.FormValue("client_id"))
	id := makeIDToken(user, r.FormValue("client_id"), "", time.Now().Unix())
	rt := randomToken()
	mu.Lock()
	refreshStore[rt] = &refreshEntry{ClientID: r.FormValue("client_id"), Username: stored.Username, Scope: stored.Scope, Expires: time.Now().Add(time.Duration(refreshTTL) * time.Second)}
	mu.Unlock()
	sendJSON(w, 200, map[string]any{"access_token": at, "token_type": "Bearer", "expires_in": accessTTL, "refresh_token": rt, "id_token": id, "scope": stored.Scope})
}

func userinfoHandler(w http.ResponseWriter, r *http.Request) {
	auth := r.Header.Get("Authorization")
	if !strings.HasPrefix(auth, "Bearer ") {
		sendJSON(w, 401, map[string]string{"error": "missing_token"})
		return
	}
	token, err := jwt.Parse(strings.TrimPrefix(auth, "Bearer "), func(t *jwt.Token) (any, error) {
		if _, ok := t.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("unexpected method")
		}
		return rsaPublicKey, nil
	})
	if err != nil {
		sendJSON(w, 401, map[string]string{"error": "invalid_token"})
		return
	}
	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok || !token.Valid {
		sendJSON(w, 401, map[string]string{"error": "invalid_token"})
		return
	}
	sub, _ := claims["sub"].(string)
	scope, _ := claims["scope"].(string)
	for _, u := range users {
		if u.Sub == sub {
			sendJSON(w, 200, extractScopeClaims(scope, u))
			return
		}
	}
	sendJSON(w, 404, map[string]string{"error": "user_not_found"})
}

func discoveryHandler(w http.ResponseWriter, r *http.Request) {
	sendJSON(w, 200, map[string]any{
		"issuer":                               issuer,
		"authorization_endpoint":               issuer + "/authorize",
		"token_endpoint":                       issuer + "/token",
		"userinfo_endpoint":                    issuer + "/userinfo",
		"jwks_uri":                             issuer + "/.well-known/jwks.json",
		"scopes_supported":                     []string{"openid", "profile", "email"},
		"response_types_supported":             []string{"code"},
		"grant_types_supported":                []string{"authorization_code", "refresh_token"},
		"id_token_signing_alg_values_supported": []string{"RS256"},
	})
}

func jwksHandler(w http.ResponseWriter, r *http.Request) {
	n := base64.RawURLEncoding.EncodeToString(rsaPublicKey.N.Bytes())
	e := base64.RawURLEncoding.EncodeToString([]byte{byte(rsaPublicKey.E >> 16), byte(rsaPublicKey.E >> 8), byte(rsaPublicKey.E)})
	sendJSON(w, 200, map[string]any{
		"keys": []map[string]string{{
			"kty": "RSA", "use": "sig", "alg": "RS256", "kid": "oidc-rsa-1", "n": n, "e": e,
		}},
	})
}

func main() {
	mux := http.NewServeMux()
	mux.HandleFunc("/authorize", authorizeHandler)
	mux.HandleFunc("/consent", consentHandler)
	mux.HandleFunc("/token", tokenHandler)
	mux.HandleFunc("/userinfo", userinfoHandler)
	mux.HandleFunc("/.well-known/openid-configuration", discoveryHandler)
	mux.HandleFunc("/.well-known/jwks.json", jwksHandler)

	addr := fmt.Sprintf("0.0.0.0:%s", getEnv("PORT", "8000"))
	log.Printf("OIDC Provider at http://localhost:%s", getEnv("PORT", "8000"))
	log.Fatal(http.ListenAndServe(addr, mux))
}

var _ = sha256.New
